spatialdata/_core/transform.py

```diff
@@ -1,86 +1,526 @@
+        return Scale(scale=new_scale)
+
+    def to_affine(self) -> Affine:
+        assert self.ndim == 2
+        m: ArrayLike = np.vstack((np.hstack((np.diag(self.scale), np.zeros((2, 1)))), np.array([0.0, 0.0, 1.0])))
+        return Affine(affine=m)
 
-        if scale_factors is None:
-            if translation is not None:
-                scale_factors = np.ones(len(translation), dtype=float)
-            elif ndim is not None:
-                scale_factors = np.ones(ndim, dtype=float)
+
+class Affine(BaseTransformation):
+    def __init__(
+        self,
+        affine: Optional[Union[ArrayLike, List[Any]]] = None,
+    ) -> None:
+        """
+        class for storing scale transformations.
+        """
+        if affine is None:
+            affine = np.array([[1, 0, 0], [0, 1, 0], [0, 0, 1]], dtype=float)
+        else:
+            if isinstance(affine, list):
+                affine = np.array(affine)
+            if len(affine.shape) == 1:
+                assert len(affine) == 6
+                affine = np.vstack([affine.reshape((2, 3)), np.array([0.0, 0.0, 1.0])])
+            assert affine.shape == (3, 3)
+
+        self.affine = affine
+        self._ndim = 2
+
+    @property
+    def ndim(self) -> Optional[int]:
+        return self._ndim
+
+    def to_dict(self) -> Dict[str, Any]:
+        return {
+            "type": "affine",
+            "affine": self.affine[:2, :].ravel().tolist(),
+        }
+
+    def transform_points(self, points: ArrayLike) -> ArrayLike:
+        p = np.vstack([points.T, np.ones(points.shape[0])])
+        q = self.affine @ p
+        return q[:2, :].T  # type: ignore[no-any-return]
+
+    def inverse(self) -> BaseTransformation:
+        # naive check for numerical stability
+        a = self.affine[0, 0]
+        b = self.affine[0, 1]
+        m = self.affine[0, 2]
+        c = self.affine[1, 0]
+        d = self.affine[1, 1]
+        n = self.affine[1, 2]
+        det = a * d - b * c
+        closed_form = np.array([[d, -c, 0], [-b, a, 0], [b * n - d * m, c * m - a * n, det]])
+        return Affine(affine=closed_form)
+
+    def to_affine(self) -> Affine:
+        return copy.deepcopy(self)
+
+
+class Rotation(BaseTransformation):
+    def __init__(
+        self,
+        rotation: Optional[Union[ArrayLike, List[Any]]] = None,
+    ) -> None:
+        """
+        class for storing scale transformations.
+        """
+        if rotation is None:
+            rotation = np.array([[1, 0], [0, 1]], dtype=float)
+        else:
+            if isinstance(rotation, list):
+                rotation = np.array(rotation)
+            if len(rotation.shape) == 1:
+                assert len(rotation) == 4
+                rotation = rotation.reshape((2, 2))
+            assert rotation.shape == (2, 2)
+
+        self.rotation = rotation
+        self._ndim = 2
+
+    @property
+    def ndim(self) -> Optional[int]:
+        return self._ndim
+
+    def to_dict(self) -> Dict[str, Any]:
```

Comment:
I'm getting more convinced that it makes sense to type dictionaries.

Reply:
Agree, I'll do it.

Reply:
Follow up on this, I tried doing this by introducing the (ugly)
```
Transformation_t = Dict[str, Union[str, List[int], List[str], List[Dict[str, Any]]]]
```
and replacing all the typing annotation with `Transformation_t`, but I have just realized that this is suboptimal since there is still a `Any` inside, and also this was giving a ton of weird mypy errors. The problem here is that `Sequence` and `MapAxis` transformations can contain infinitely nested transformations. Not sure if there is a way around this to support for these recursive dictionary types.

Meanwhile I have kept `Transformation_t` around the code, but I have redefined it as `Dict[str, Any]`.

Reply:
I would maybe keep it like this and eventually if we want to give a better typing I would open a relative issue for the future.